src/service/utils/download.py

```diff
@@ -0,0 +1,172 @@
+    elif column_name == "trustyai.INDEX":
+        if errors:
+            combined_error = ", ".join(errors)
+            raise HTTPException(status_code=400, detail=combined_error)
+        min_val, max_val = sorted([int(v) for v in values])
+        mask = (df[column_name] >= min_val) & (df[column_name] < max_val)
+    else:
+        if not all(isinstance(v, (int, float)) for v in values):
+            errors.append(
+                "BETWEEN operation must only contain numbers, describing the lower and upper bounds of the desired range. Received non-numeric values"
+            )
+        if errors:
+            combined_error = ", ".join(errors)
+            raise HTTPException(status_code=400, detail=combined_error)
+        min_val, max_val = sorted(values)
+        try:
+            mask = (pd.to_numeric(df[column_name], errors="coerce") >= min_val) & (
+                pd.to_numeric(df[column_name], errors="coerce") < max_val
+            )
+        except:
+            mask = (df[column_name].astype(str) >= str(min_val)) & (df[column_name].astype(str) < str(max_val))
+    if negate:
+        mask = ~mask
+    return df[mask]
+
+
+async def load_model_dataframe(model_id: str) -> pd.DataFrame:
+    """Load model dataframe from storage."""
+    storage = get_storage()
+    try:
+        input_data, input_cols = await storage.read_data(f"{model_id}_inputs")
+        output_data, output_cols = await storage.read_data(f"{model_id}_outputs")
+        metadata_data, metadata_cols = await storage.read_data(f"{model_id}_metadata")
+        if input_data is None or output_data is None or metadata_data is None:
+            raise HTTPException(status_code=404, detail=f"Model {model_id} not found")
+        df = pd.DataFrame()
+        if len(input_data) > 0:
+            if input_data.ndim == 2 and len(input_cols) == 1 and input_data.shape[1] > 1:
+                col_name = input_cols[0]
+                for j in range(input_data.shape[1]):
+                    df[f"{col_name}_{j}"] = input_data[:, j]
+            else:
+                input_df = pd.DataFrame(input_data, columns=input_cols)
+                for col in input_cols:
+                    df[col] = input_df[col]
+        if len(output_data) > 0:
+            if output_data.ndim == 2 and len(output_cols) == 1 and output_data.shape[1] > 1:
+                col_name = output_cols[0]
+                for j in range(output_data.shape[1]):
+                    df[f"{col_name}_{j}"] = output_data[:, j]
+            else:
+                if output_data.ndim == 2:
+                    output_data = output_data.flatten()
+                output_df = pd.DataFrame({output_cols[0]: output_data})
+                for col in output_cols:
+                    df[col] = output_df[col]
+        if len(metadata_data) > 0 and isinstance(metadata_data[0], bytes):
+            deserialized_metadata = []
+            for row in metadata_data:
+                deserialized_row = pickle.loads(row)
```

Comment:
unsure why this is necessary- is the metadata returned from storage.read_data  ever serialized? If so, that's a bug in the storage code